kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -249,38 +283,126 @@ private static boolean allowMappingModeChange(String oldMode, String newMode) {
      */
     private static Metadata assignColumnIdAndPhysicalName(Metadata metadata, boolean isNewTable) {
         StructType schema = metadata.getSchema();
-        int maxColumnId = Math.max(
+        AtomicInteger maxColumnId = new AtomicInteger(
+                Math.max(
                 Integer.parseInt(metadata.getConfiguration()
                         .getOrDefault(COLUMN_MAPPING_MAX_COLUMN_ID_KEY, "0")),
-                findMaxColumnId(schema));
+                findMaxColumnId(schema)));
         StructType newSchema = new StructType();
         for (StructField field : schema.fields()) {
-            StructField updatedField = field;
-
-            if (!hasColumnId(field)) {
-                maxColumnId++;
-                updatedField = field.withNewMetadata(FieldMetadata.builder()
-                        .fromMetadata(field.getMetadata())
-                        .putLong(COLUMN_MAPPING_ID_KEY, maxColumnId)
-                        .build());
-            }
-            if (!hasPhysicalName(field)) {
-                // re-use old display names as physical names when a table is updated
-                String physicalName = isNewTable ? "col-" + UUID.randomUUID() : field.getName();
-                updatedField = updatedField.withNewMetadata(FieldMetadata.builder()
-                        .fromMetadata(updatedField.getMetadata())
-                        .putString(COLUMN_MAPPING_PHYSICAL_NAME_KEY, physicalName)
-                        .build());
-            }
-            newSchema = newSchema.add(updatedField);
+            newSchema = newSchema.add(
+                    transformSchema(
+                            assignColumnIdAndPhysicalName(field, maxColumnId, isNewTable),
+                            maxColumnId,
+                            isNewTable));
+        }
+
+        if (Boolean.parseBoolean(metadata.getConfiguration()
```

Comment:
Can't you do: `TableConfig.ICEBERG_COMPAT_V2_ENABLED.fromMetadata(engine, metadata)`

That seems to require the engine ... if even we cannot be bothered to correctly use the `.fromMetadata` helper method then why would our connector developers be bothered to use it?

i.e. this seems to require the `engine` be passed to a bunch of these interfaces. Is there a way around this? cc @vkorukanti

Reply:
yeah it would be great to use `TableConfig.ICEBERG_COMPAT_V2_ENABLED.fromMetadata(..)` here but I don't know why the `engine` is required there for engine-specific validation, so I've opted to check the `ICEBERG_COMPAT_V2_ENABLED` in a different way